### PR TITLE
fix(50303): Using @linkcode in combination with private methods causes TS1003 in JavaScript files.

### LINKED ISSUE #50303: Using `@linkcode` in combination with private methods causes TS1003 in JavaScript files.
# Bug Report

### 🔎 Search Terms

linkcode private 

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about `linkcode`

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.8.0-beta&filetype=js#code/MYGwhgzhAEBiD29oG8BQ0PQPQCofs2h2gBUALASxgHd4AnAawgBpoAXM+AVwHMz2yAU2ggKAOwbQAJvEEQxAcjbQIgwQFt28ApmK1G0cQOEAHcAE8edbmKkAuHRmLIAAqInB4U4RyoA6MXg2AAU6CgA3MDZBAFlBDi8AX0ciaBTicipoHgi5aDAxaEE6azoHQlTXdwZPbwF-AGJw4vNQiKjY+M4pZIqcLEcIeHVOhKkACgBKFGhkx0CQsMjouLGpmbnCJpa25dHu9eRZ1ESgA)

### 💻 Code

```js
class Foo {
    /**
     * This works, though the link doesn't seem to
     * work in the playground:
     * {@linkcode this.notPrivateMethod}
     * 
     * This gives an error:
     * {@linkcode this.#veryPrivateMethod}
     */
    someMethod() { }

    notPrivateMethod() { }

    #veryPrivateMethod() { }
}
```

### 🙁 Actual behavior

```
Identifier expected. TS1003
```

### 🙂 Expected behavior

No error

### PR DESCRIPTION
**Pull request by a-tarasyuk**
Fixes #50303

### CODE REVIEW

File: src/compiler/parser.ts [hunk 1]
```diff
@@ -9226,6 +9215,24 @@ namespace Parser {
                 return finishNode(create(name, text.join("")), start, scanner.getTokenEnd());
             }
 
+            function parseJSDocLinkName() {
+                if (tokenIsIdentifierOrKeyword(token())) {
+                    const pos = getNodePos();
+
+                    let name: EntityName | JSDocMemberName = parseIdentifierName();
+                    while (parseOptional(SyntaxKind.DotToken)) {
+                        name = finishNode(factory.createQualifiedName(name, token() === SyntaxKind.PrivateIdentifier ? createMissingNode<Identifier>(SyntaxKind.Identifier, /*reportAtCurrentPosition*/ false) : parseIdentifier()), pos);
```
Comment (sandersn) [inline src/compiler/parser.ts:9224]: the key difference in the new code is that it's a choice between parseIdentifier and createMissingNode. In the old code, parseRightSideOfDot chooses between five different functions. That seems better, but obviously it was wrong. How was it wrong?

**Review by sandersn — CHANGES_REQUESTED** [review]
I don't understand why this fix works. Can you explain?
Also I'd like a couple more tests.

File: src/compiler/parser.ts (hunk 1 above)
Comment (a-tarasyuk) [reply src/compiler/parser.ts:9224]: @sandersn Thanks for the review. `parseRightSideOfDot` always throws an error for private identifiers that should be allowed in this case. (`/*reportAtCurrentPosition*/ true, Diagnostics.Identifier_expected`). There is an option to control this with a new parameter, which requires a deep function chain change, or just use `createMissingNode` directly

https://github.com/microsoft/TypeScript/blob/5f5ed0f487d43127257154e999eefc1f25208368/src/compiler/parser.ts#L3611-L3613

https://github.com/microsoft/TypeScript/blob/5f5ed0f487d43127257154e999eefc1f25208368/src/compiler/parser.ts#L2598-L2603

File: src/compiler/parser.ts [hunk 2]
```diff
@@ -9203,18 +9203,7 @@ namespace Parser {
                 }
                 nextTokenJSDoc(); // start at token after link, then skip any whitespace
                 skipWhitespace();
-                // parseEntityName logs an error for non-identifier, so create a MissingNode ourselves to avoid the error
```
[the remark this answers is not in the document]
Comment (a-tarasyuk) [reply src/compiler/parser.ts:9206]: I've added more details about this issue and this comment seems to be false as `createMissingNode` is also logging the error...

File: tests/cases/conformance/jsdoc/jsdocLinkTag7.ts [hunk 3]
```diff
@@ -0,0 +1,14 @@
+// @checkJs: true
+// @allowJs: true
+// @target: esnext
+// @noEmit: true
+// @filename: /a.js
+class Foo {
+    /**
+     * {@linkcode this.a}
+     * {@linkcode this.#c}
```
[the remark this answers is not in the document]
Comment (a-tarasyuk) [reply tests/cases/conformance/jsdoc/jsdocLinkTag7.ts:9]: I've added.

**Review by sandersn — APPROVED** [review]

**Pull request merged**